### ISSUE TITLE
Performance degradation in 0.8.0-rc4

### ISSUE THREAD

**Issue by kenota** [dropwizard/dropwizard#907]
I discovered that simple app using hibernate degrades in performance from 12k req/sec to 2.2k req/sec when switching dropwizard version from 0.7.1 to 0.8.0-rc4.

Steps to reproduce:

1. `git clone git@github.com:kenota/dropwizard-perf-issue.git`
2. `cd dropwizard-perf-issue`
4. `gradle singleJar`
5. `java -jar ./build/libs/app.jar server hello-world.yml  > /dev/null`

At this point you should run some benchmark, against http://localhost:8080/video-hibernate URL. I was able to achieve 12k req/sec using 50 threads on my MacbookPro.

6. Ctrl+C to stop the running app
7. Uncomment `dropwizardVersion = "0.8.0-rc4"` line in `build.gradle` and comment out `dropwizardVersion = "0.7.1"`
8. `gradle singleJar`
9. `java -jar ./build/libs/app.jar server hello-world.yml  > /dev/null`

Run benchmark again. At this point, I was able to get only 2.2k req/sec using same 50 threads.

[Repo with code showing problem](https://github.com/kenota/dropwizard-perf-issue)

**Comment by devinrsmith** [comment]
I ran into a potentially similar issue a couple months ago with 0.7.1 vs 0.8.0-rc1, but didn't have the time to verify what I was seeing. This sounds very similar.

**Comment by arteam** [comment]
@kenota Thanks very much for the report. We are glad to hear that people upgrade to the release candidate builds and report issues they have encountered. It helps to us to make Dropwizard better.

About the issue, I can confirm about 6x perfomance degradation in 0.8.0-rc4.
I've got the following results in my test environment: 
* (two modest linux virtual machines: one for the server, one for the client)
* Java(TM) SE Runtime Environment (build 1.8.0_20-b26)
* JVM arguments `-server -Xmx512m -Xms512m`
* Standard wrk test
* 5 tests for 30 seconds. First 3 are ignored.

0.7.1 
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.27ms    3.74ms  39.28ms   87.46%
    Req/Sec     2.47k     1.02k    5.67k    68.50%
  141398 requests in 30.00s, 28.99MB read
Requests/sec:   4713.28
Transfer/sec:      0.97MB
````

0.8.0-rc4
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.47ms    7.75ms  62.17ms   62.78%
    Req/Sec   354.07     53.65   510.00     68.61%
  21163 requests in 30.00s, 3.49MB read
Requests/sec:    705.39
Transfer/sec:    119.17KB
````

I will try to make some investigation of this behaviour.

**Comment by JonMR** [comment]
Is Dropwizard at a point where performance tests are necessary? Setting up infrastructure for performance tests is not a trivial task but may catch these degradations earlier.




I'd be happy to take the discussion to the mailing list.



—

**Comment by arteam** [comment]
At first glance it looks like a some _dropwizard-hibernate_ integration issue. Numbers look really weird.

0.7.1 Hibernate stats
````
INFO  [2015-02-28 09:55:31,358] org.hibernate.engine.internal.StatisticalLoggingSessionEventListener: Session Metrics {
    2925 nanoseconds spent acquiring 1 JDBC connections;
    0 nanoseconds spent releasing 0 JDBC connections;
    3348 nanoseconds spent preparing 1 JDBC statements;
    3427 nanoseconds spent executing 1 JDBC statements;
    0 nanoseconds spent executing 0 JDBC batches;
    0 nanoseconds spent performing 0 L2C puts;
    0 nanoseconds spent performing 0 L2C hits;
    0 nanoseconds spent performing 0 L2C misses;
    12915 nanoseconds spent executing 1 flushes (flushing a total of 2 entities and 0 collections);
    1117 nanoseconds spent executing 1 partial-flushes (flushing a total of 0 entities and 0 collections)
}
````
0.7.1 **findAll** method timer
````json
{
    "findAll": {
      "count" : 184286,
      "max" : 0.004406585,
      "mean" : 8.251789688715953E-5,
      "min" : 3.76E-5,
      "p50" : 6.41035E-5,
      "p75" : 7.735575E-5,
      "p95" : 1.0673735E-4,
      "p98" : 1.3669447999999998E-4,
      "p99" : 2.068743000000001E-4,
      "p999" : 0.0043515032120000074,
      "stddev" : 1.887684948495728E-4,
      "m15_rate" : 458.3885218384127,
      "m1_rate" : 1262.7595698455427,
      "m5_rate" : 499.6565012876884,
      "mean_rate" : 412.5573303707618,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    }
}
}
````

0.8.0-rc4 Hibernate stats
````
INFO  [2015-02-28 10:05:10,813] org.hibernate.engine.internal.StatisticalLoggingSessionEventListener: Session Metrics {
    5102 nanoseconds spent acquiring 1 JDBC connections;
    0 nanoseconds spent releasing 0 JDBC connections;
    5052 nanoseconds spent preparing 1 JDBC statements;
    9358 nanoseconds spent executing 1 JDBC statements;
    0 nanoseconds spent executing 0 JDBC batches;
    0 nanoseconds spent performing 0 L2C puts;
    0 nanoseconds spent performing 0 L2C hits;
    0 nanoseconds spent performing 0 L2C misses;
    17067 nanoseconds spent executing 1 flushes (flushing a total of 2 entities and 0 collections);
    1095 nanoseconds spent executing 1 partial-flushes (flushing a total of 0 entities and 0 collections)
}
````

0.8.0-rc4  **findAll** method timer
````json
{
    "findAll": {
        "count": 111399,
        "max": 0.076793583,
        "mean": 0.012456742886687444,
        "min": 0.001129987,
        "p50": 0.010994974000000001,
        "p75": 0.019294024,
        "p95": 0.029016877000000003,
        "p98": 0.033430958000000004,
        "p99": 0.036311203,
        "p999": 0.059136342,
        "stddev": 0.00946485908806412,
        "m15_rate": 233.84245379553187,
        "m1_rate": 491.7529578715853,
        "m5_rate": 291.77870315313703,
        "mean_rate": 323.76017533370344,
        "duration_units": "seconds",
        "rate_units": "calls/second"
    }
}
````

The drop is fierce. From 82 microseconds to 12 **milli**seconds.

From the metrics I can conclude that the database and the connection pool behaves roughly the same and return expected numbers for the in-process database. Young GC numbers also look similar, apart from the fact that 0.7.1 generates a lot more garbage and GC starting become a bottleneck and throws FullGCs.

0.7.1
````
{
    "jvm.gc.PS-Scavenge.count": {
        "value": 95
    },
    "jvm.gc.PS-Scavenge.time": {
        "value": 4240
    },
    "jvm.gc.PS-MarkSweep.count": {
        "value": 46
    },
    "jvm.gc.PS-MarkSweep.time": {
        "value": 30523
    }
}
````
0.8.0-rc4
````
{
    "jvm.gc.PS-MarkSweep.count": {
        "value": 1
    },
    "jvm.gc.PS-MarkSweep.time": {
        "value": 178
    },
    "jvm.gc.PS-Scavenge.count": {
        "value": 107
    },
    "jvm.gc.PS-Scavenge.time": {
        "value": 1654
    }
}
````
This means there is a bottleneck somewhere between the application and the database pool.

For a note 0.7.1 uses _Hibernate Core 4.3.5.Final_ and 0.8.0-rc4 uses _Hibernate Core 4.3.8.Final_

**Comment by arteam** [comment]
By the way, there is no regression with JDBI (little worse througput, but better latency).

0.7.1
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate/jdbi
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.41ms   10.80ms  54.62ms   93.25%
    Req/Sec     2.64k     1.31k    6.67k    66.69%
  150112 requests in 30.00s, 30.78MB read
Requests/sec:   5003.73
Transfer/sec:      1.03MB
````

0.8.0-rc4
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate/jdbi
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.91ms    2.98ms  32.50ms   89.06%
    Req/Sec     2.26k   733.29     4.22k    74.94%
  130637 requests in 30.00s, 21.55MB read
Requests/sec:   4354.40
Transfer/sec:    735.66KB
````

**Comment by arteam** [comment]
Well, it actually looks like a Hibernate issue. I couln'd find any changes in `dropwizard-hibernate`, 
apart from upgrading to newer Hibernate versions that could affect perfomance so drastically.

Running 0.8.0-rc5-SNAPSHOT with Hibernate Core `4.3.5.Final` gave me similar results to 0.7.1 
(worse but acceptable)
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     4.02ms    3.87ms  40.60ms   88.22%
    Req/Sec     1.72k   627.90     3.60k    70.15%
  98185 requests in 30.00s, 16.20MB read
Requests/sec:   3272.84
Transfer/sec:    552.93KB
````
````json
{
  "findAll" : {
      "count" : 690158,
      "max" : 0.018245754,
      "mean" : 1.1480092781416619E-4,
      "min" : 4.0662000000000005E-5,
      "p50" : 7.0688E-5,
      "p75" : 8.1925E-5,
      "p95" : 1.1272300000000001E-4,
      "p98" : 1.39163E-4,
      "p99" : 1.55498E-4,
      "p999" : 0.018245754,
      "stddev" : 6.758212715369515E-4,
      "m15_rate" : 880.4983235729563,
      "m1_rate" : 1083.995873656406,
      "m5_rate" : 1358.0522535970188,
      "mean_rate" : 1732.7668884405502,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    }
}
````

And the same application with 4.3.6-Final
````
Running 30s test @ http://192.168.52.15:18080/video-hibernate
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.26ms    9.78ms 107.27ms   72.97%
    Req/Sec   325.21     50.95   476.00     72.36%
  19362 requests in 30.00s, 3.19MB read
Requests/sec:    645.34
Transfer/sec:    109.03KB
````

I've revieved changes in Hibernate 4.3.6 [release notes](https://hibernate.atlassian.net/secure/ReleaseNote.jspa?projectId=10031&version=16550), but didn't find any changes related to perfomance. 

For the time being, I consider to rollback Dropwizard to Hibernate 4.3.5.

**Comment by arteam** [comment]
I've also checked this against a dedicated Postgres database and got some performance degradation.
Not so drastic, but considerable (2x-3x).
It would be great if someone could also make some load tests on real data and real hardware and 
share his/her experience.

**Comment by arteam** [comment]
@kenota Could you, please, test against the last 0.8.0-rc5-SNAPSHOT? It should be available in the Sonatype Snapshots repository in a couple of hours.

**Comment by arteam** [comment]
@JonMR Performance testing is a big separate issue. We don't have any benchmarks (micro or macro) yet. Currently we rely to the developers of the libraries we use and trust them. Our strategy is try to update your dependencies to the latest versions, which usually bring performance improvements.
To all appearances, sometimes this does not work.

Definitely there is a room to improve.

**Comment by carlo-rtr** [comment]
0.8.0-rc5 was just published, should be on sonatype central releases momentarily

**Comment by kenota** [comment]
@arteam, thank you for suck quick reaction, especially on Saturday.  I have tried my code with 0.8.0-rc5 and can confirm that issue seems to be resolved with 0.8.0-rc5. 

Should I close this issue myself or somebody from dropwizard project should do that?

**Comment by arteam** [comment]
Glad to hear that the problem is resolved. I will close the issue and link it with the 0.8 release.

Thanks again for catching and reporting this.
